server/server_test.go

```diff
@@ -274,6 +276,121 @@ func TestSubmitCertificateVerification(t *testing.T) {
 	require.NoError(t, eg.Wait())
 }
 
+func createTestKeyFile(t *testing.T, dir, name string, data []byte) string {
+	path := filepath.Join(dir, name)
+	if err := os.WriteFile(path, data, 0o644); err != nil {
+		t.Fatalf("failed to write key file: %v", err)
+	}
+	return path
+}
+
+// Test submitting a challenge followed by proof generation and getting the proof via GRPC.
+func TestLoadTrustedKeysAndSubmit(t *testing.T) {
+	const keysNum = 2
+
+	// create test files
+	dir := t.TempDir()
+	t.Cleanup(func() { os.RemoveAll(dir) })
+
+	var (
+		trustedKeyCert *shared.OpaqueCert
+		trustedKey     []byte
+	)
+
+	challenge := []byte("challenge")
+
+	// User credentials
+	userPubKey, userPrivKey, err := ed25519.GenerateKey(rand.Reader)
+	require.NoError(t, err)
+
+	signature := ed25519.Sign(userPrivKey, challenge)
+
+	for i := 0; i < keysNum; i++ {
+		pubKey, private, err := ed25519.GenerateKey(nil)
+		require.NoError(t, err)
+
+		createTestKeyFile(t, dir, fmt.Sprintf("valid_key_%d.key", i), []byte(base64.StdEncoding.EncodeToString(pubKey)))
+
+		if i == keysNum-1 {
+			expiration := time.Now().Add(time.Hour)
+			data, err := shared.EncodeCert(&shared.Cert{Pubkey: userPubKey, Expiration: &expiration})
+			require.NoError(t, err)
+
+			trustedKeyCert = &shared.OpaqueCert{
+				Data:      data,
+				Signature: ed25519.Sign(private, data),
+			}
+			trustedKey = pubKey
+		}
+	}
```

Comment:
NIT: This can be simplified:

```suggestion
	// generate keys
	for i := range keysNum - 1 {
		pubKey, _, err := ed25519.GenerateKey(nil)
		require.NoError(t, err)

		path := filepath.Join(dir, fmt.Sprintf("valid_key_%d.key", i))
		require.NoError(t, os.WriteFile(path, []byte(base64.StdEncoding.EncodeToString(pubKey)), 0o644))
	}
	// generate key and certificate
	trustedKey, private, err := ed25519.GenerateKey(nil)
	require.NoError(t, err)

	path := filepath.Join(dir, "valid_key.key")
	require.NoError(t, os.WriteFile(path, []byte(base64.StdEncoding.EncodeToString(trustedKey)), 0o644))
	expiration := time.Now().Add(time.Hour)
	data, err := shared.EncodeCert(&shared.Cert{Pubkey: userPubKey, Expiration: &expiration})
	require.NoError(t, err)

	trustedKeyCert = &shared.OpaqueCert{
		Data:      data,
		Signature: ed25519.Sign(private, data),
	}
```